File: bindings/ruby/test/test_cross_project.rb
require "minitest/autorun"
require "tmpdir"
require "set"
require "vinary_tree/libdictenstein"
require "vinary_tree/liblevenshtein"

class CrossProjectTest < Minitest::Test
  LD = VinaryTree::Libdictenstein
  LL = VinaryTree::Liblevenshtein

  def collect(query) = query.map(&:term).to_set

  def test_long_lived_query_start_snapshot_and_property_traces
    dictionary = LD::DynamicDawg.new
    dictionary.put_all([["cat", 1], ["cot", 2], ["cut", 3], ["scat", nil]])
    transducer = LL::Transducer.new(dictionary)
    enumerator = transducer.query("cat", 2).each
    old = Set[enumerator.next.term]
    dictionary.remove("cot"); dictionary.put("cut", 30); dictionary.put("cit", 5); dictionary.compact
    begin
      old << enumerator.next.term while true
    rescue StopIteration
      nil
    end
    assert_equal Set["cat", "cot", "cut", "scat"], old
    64.times do |seed|
      subject = LD::DynamicDawg.new
      32.times { |index| subject.put(format("t%02d", index), index) }
      machine = LL::Transducer.new(subject)
      expected = collect(machine.query("t00", 3))
      stream = machine.query("t00", 3).each
      actual = Set[stream.next.term]
      subject.remove(format("t%02d", seed % 32)); subject.put(format("x%02d", seed), seed)
      begin
        actual << stream.next.term while true
      rescue StopIteration
        nil
      end
      assert_equal expected, actual, "seed #{seed}"
      machine.close; subject.close
    end
  ensure
    transducer&.close; dictionary&.close
  end

  def test_other_backends
    dat = LD::DoubleArrayTrie.new([["alpha", 7], ["beta", nil]])
    assert_equal 7, dat.get("alpha").value
    suffix = LD::Scdawg.new; suffix.put("banana", 1)
    assert suffix.include_substring?("ana"); assert_equal 2, suffix.substring_frequency("ana")
    Dir.mktmpdir do |directory|
      path = File.join(directory, "dictionary")
      persistent = LD::PersistentArtrie.create(path); persistent.put("durable", 9); persistent.checkpoint; persistent.close
      reopened = LD::PersistentArtrie.open(path); assert_equal 9, reopened.get("durable").value; reopened.close
    end
  ensure
    dat&.close; suffix&.close
  end
end

File: bindings/python/tests/test_cross_project.py
from __future__ import annotations

from pathlib import Path

import libdictenstein
import liblevenshtein


def test_crud_batch_and_cross_project_query_snapshot() -> None:
    dictionary = libdictenstein.DynamicDawg()
    assert (
        dictionary.update_many([("cat", 1), ("cot", 2), ("cut", 3), ("scat", None)])
        == 4
    )
    assert len(dictionary) == 4
    assert dictionary.lookup("cot") == (True, 2)

    automaton = liblevenshtein.Transducer(dictionary)
    frozen = sorted(automaton.query("cat", 2), key=lambda item: str(item.term))
    cursor = automaton.query("cat", 2)
    first = next(cursor)

    assert dictionary.remove("cot")
    assert not dictionary.insert("cut", 30)
    assert dictionary.insert("cit", 5)
    dictionary.compact()
    dictionary.clear()
    assert dictionary.insert("new", 99)

    assert [item.term for item in automaton.query("cat", 8)] == ["new"]
    automaton.close()
    dictionary.close()

    observed = sorted([first, *cursor], key=lambda item: str(item.term))
    assert observed == frozen


def test_double_array_trie_composes_without_copying() -> None:
    dictionary = libdictenstein.DoubleArrayTrie(
        [("café", 7), ("caff", None), ("tea", 9)]
    )
    automaton = liblevenshtein.Transducer(dictionary)
    observed = sorted(automaton.query("cafe", 2), key=lambda match: str(match.term))
    assert [(match.term, match.id) for match in observed] == [
        ("caff", None),
        ("café", 7),
    ]
    automaton.close()
    dictionary.close()


def test_scdawg_long_lived_cursor_keeps_query_start_revision() -> None:
    dictionary = libdictenstein.Scdawg()
    assert dictionary.update_many([("cat", 1), ("cot", 2), ("cut", None)]) == 3
    automaton = liblevenshtein.Transducer(dictionary)
    cursor = automaton.query("cat", 2)
    first = next(cursor)
    assert dictionary.insert("cit", 4)
    frozen = sorted([first, *cursor], key=lambda match: str(match.term))
    assert "cit" not in {match.term for match in frozen}
    automaton.close()
    dictionary.close()


def test_persistent_artrie_checkpoint_keeps_query_start_revision(
    tmp_path: Path,
) -> None:
    path = tmp_path / "terms.part"
    dictionary = libdictenstein.PersistentARTrie.create(path)
    assert dictionary.update_many([("cat", 1), ("cot", 2), ("cut", None)]) == 3
    dictionary.checkpoint()
    automaton = liblevenshtein.Transducer(dictionary)
    cursor = automaton.query("cat", 2)
    first = next(cursor)
    assert dictionary.remove("cot")
    assert not dictionary.insert("cut", 30)
    assert dictionary.insert("cit", 4)
    dictionary.checkpoint()
    frozen = sorted([first, *cursor], key=lambda match: str(match.term))
    assert [(match.term, match.id) for match in frozen] == [
        ("cat", 1),
        ("cot", 2),
        ("cut", None),
    ]
    automaton.close()
    dictionary.close()
